core/src/main/kotlin/com/strumenta/kolasu/ids/CachingNodeIDProvider.kt

```diff
@@ -0,0 +1,29 @@
+package com.strumenta.kolasu.ids
+
+import com.strumenta.kolasu.model.Node
+import java.util.IdentityHashMap
+
+/**
+ * The Node ID of most nodes is a Positional Node ID. Positional Node IDs are calculated by considering the Node ID of
+ * the parent,so that the Node IDs of all ancestors are calculated again and again.
+ * By using a cache we can avoid that.
+ */
+class CachingNodeIDProvider(val wrapped: NodeIdProvider) : BaseNodeIdProvider() {
```

Comment:
I've learnt that whenever I see a cache, without a method to clear it, my spider sense goes off..

Reply:
Good point. This is intended to be short-lived but I will add a method to clear it.

The idea is that when traversing a tree:
A
|          \
B1       B2
|    \       |     \
C1 C2  C3 C4

We will end up calculating the node ID of A 7 times (once to set the node ID of A, then when setting the one of the other six nodes, that all depend on the node ID of A). So the idea was to use a CachingNodeIdProvider instance for processing an AST and then throw it away.